Add EIP-6122: Forkid checks based on timestamps (#6122)
ethereum/EIPs: EIPS/eip-6122.md

## Specification

Each node maintains the following values:
- **`FORK_HASH`**: IEEE CRC32 checksum (`[4]byte`) of the genesis hash and fork blocks numbers or timestamps that already passed.
  - The fork block numbers or timestamps are fed into the CRC32 checksum in ascending order.
  - If multiple forks are applied at the same block or time, the block number or timestamp is checksummed only once.
  - Block numbers are regarded as `uint64` integers, encoded in big endian format when checksumming.
  - Block timestamps are regarded as `uint64` integers, encoded in big endian format when checksumming.
  - If a chain is configured to start with a non-Frontier ruleset already in its genesis, that is NOT considered a fork.
- **`FORK_NEXT`**: Block number or timestamp (`uint64`) of the next upcoming fork, or `0` if no next fork is known.

A `FORK_HASH` for a time stamp based fork at `1668000000` on top of homestead would be:
- forkhash₁ = `0xcb37b2ee` (homestead+fictional fork) = `CRC32(<genesis-hash> || uint64(1150000) || 1668000000)`

### Additional rules
The following additional rules are applied:

> - Forks by timestamp have to be scheduled at or after the forks by block

Reviewer: By whom, when? Please clarify what this refers to

Reviewer 2: there **could** be forks by block intermingled with fork by timestamps which do not need coordination with the CL hardforks and technically it could be possible to have block based hfs intermixed with timestamp based. 

But it would be better to not have this pattern, and once switched to timestamp based, only schedule timestamp based.

> - The implementation needs  to filter the incoming forkids first by block then by timestamp

Reviewer: "The implementation" is a bit vague. Which implementation? 
"An implementation of fork-id verification of remote peer [...]", or 
"An implementation of a fork-id generator must [...]"

## Backwards Compatibility

> This change modifies the forkid calculation slightly which might make nodes drop peers that have not updated to the newest version as soon as the first fork on timestamp happens. 

Reviewer: ```suggestion
This change modifies the forkid calculation. As a consequence,  as soon as timestamp-scheduled fork occurs, nodes applying this change will drop peers who are not applying this change.  
```
perhaps?

---
eip: 6122

> title: Forkid checks based on timestamps

Reviewer 3: Perhaps you meant forbid?
```suggestion
title: Forbid checks based on timestamps
```

> title: Forbid checks based on timestamps

Reviewer 3: ```suggestion
title: Forkid checks based on timestamps
```

> author: Marius van der Wijden (@MariusVanDerWijden)
> description: Modifies the forkid checks to work with timestamps and block numbers

Reviewer 4: ```suggestion
description: Modifies the forkid checks to work with timestamps and block numbers
author: Marius van der Wijden (@MariusVanDerWijden)
```

discussions-to: https://ethereum-magicians.org/t/eip-6122-forkid-checks-based-on-timestamps/12130
status: Draft
type: Standards Track
category: Networking
created: 2022-12-13
requires: 2124
---

## Abstract

> [EIP-2124](eip-2124.md) proposed a way of identifying nodes on the p2p network based on their chain configuration via the forkid parameter. 

Reviewer 5: ```suggestion
[EIP-2124](./eip-2124.md) proposed a way of identifying nodes on the p2p network based on their chain configuration via the forkid parameter. 
```

It allows nodes to cut incompatible nodes off quickly which makes the P2P network more reliable.

> After the merge forks are scheduled by blocktime instead of blocknumber. This EIP updates the Forkid calculation with blocktime.

Reviewer 6: ```suggestion
After the merge, forks are scheduled by blocktime instead of blocknumber. This EIP updates the forkid calculation with blocktime.
```

> After the merge, forks are scheduled by blocktime instead of blocknumber. This EIP updates the forkid calculation with blocktime.

Reviewer 7: ```suggestion
After the merge, forks are scheduled by block time instead of block number. This EIP updates the forkid calculation with block time.
```

## Motivation

> While the execution layer scheduled forks by blocknumber, the consensus layer scheduled forks by slot number. The slot number is a time based measurement. In order to schedule forks at the same time on the consensus and execution layer, the execution layer is forced to also schedule forks by timestamp after the merge.

Reviewer 6: The execution layer had a consensus -- proof-of-work. I think it's clearer to call that out explicitly 

```suggestion
While in proof-of-work forks were scheduled by block number, the proof-of-stake consensus layer schedules forks by slot number. The slot number is a time based measurement. In order to schedule forks at the same time on the consensus and execution layer, the execution layer is forced to also schedule forks by timestamp after the merge.
```

The forkid calculation allows peers to quickly determine the configuration of peers and disconnect peers that are misconfigured or configured for other networks.

## Specification

Each node maintains the following values:

- **`FORK_HASH`**: IEEE CRC32 checksum (`[4]byte`) of the genesis hash and fork blocks numbers or timestamps that already passed.
  - The fork block numbers or timestamps are fed into the CRC32 checksum in ascending order.
  - If multiple forks are applied at the same block or time, the block number or timestamp is checksummed only once.
  - Block numbers are regarded as `uint64` integers, encoded in big endian format when checksumming.
  - Block timestamps are regarded as `uint64` integers, encoded in big endian format when checksumming.
  - If a chain is configured to start with a non-Frontier ruleset already in its genesis, that is NOT considered a fork.

> - **`FORK_NEXT`**: Block number or timestamp (`uint64`) of the next upcoming fork, or `0` if no next fork is known.

Reviewer 5: How does one distinguish between block number and timestamps? Based on that the current timestamp is around 30 bits and we're not going to reach that many blocks?

Or there's no need to distinguish? If so, please leave a note about this.

> A `FORK_HASH` for a time stamp based fork at `1668000000` on top of homestead would be:

Reviewer 7: ```suggestion
A `FORK_HASH` for a timestamp based fork at `1668000000` on top of homestead would be:
```

- forkhash₁ = `0xcb37b2ee` (homestead+fictional fork) = `CRC32(<genesis-hash> || uint64(1150000) || uint64(1668000000))`

### Additional rules

The following additional rules are applied:

> - Forks by timestamp have to be scheduled at or after the forks by block (on mainnet as well as on private networks)

Reviewer 6: I think we should make this a stronger statement. Once a forkid has shifted to timestamp, it must not return back to block

> - An implementation of fork-id verification of remote peer needs to filter the incoming forkids first by block then by timestamp

Reviewer 6: ```suggestion
- An implementation of forkid verification of remote peer needs to filter the incoming forkids first by block then by timestamp
```

## Rationale
Shanghai will be scheduled by timestamp thus the forkid calculations need to be updated to work with timestamps and blocks.

> Since all block based forks are before time based forks, nodes need to check the block based forks before the time based forks.

Reviewer 6: ```suggestion
Since all block number based forks are before time based forks, nodes need to check the block based forks before the time based forks.
```

## Backwards Compatibility

This change modifies the forkid calculation slightly. 
As a consequence nodes applying this change will drop peers who are not applying this change as soon as timestamp-scheduled fork occurs.
This is not only expected, but actually the purpose of the forkid in the first place.

> ## Security Considerations 
> No known security risks

Reviewer 4: ```suggestion
```

		{9200000, 0, ID{Hash: checksumToBytes(0xe029e991), Next: 12244000}},    // First Muir Glacier block
		{12243999, 0, ID{Hash: checksumToBytes(0xe029e991), Next: 12244000}},   // Last Muir Glacier block
		{12244000, 0, ID{Hash: checksumToBytes(0x0eb440f6), Next: 12965000}},   // First Berlin block
		{12964999, 0, ID{Hash: checksumToBytes(0x0eb440f6), Next: 12965000}},   // Last Berlin block
		{12965000, 0, ID{Hash: checksumToBytes(0xb715077d), Next: 13773000}},   // First London block
		{13772999, 0, ID{Hash: checksumToBytes(0xb715077d), Next: 13773000}},   // Last London block
		{13773000, 0, ID{Hash: checksumToBytes(0x20c327fc), Next: 15050000}},   // First Arrow Glacier block
		{15049999, 0, ID{Hash: checksumToBytes(0x20c327fc), Next: 15050000}},   // Last Arrow Glacier block
		{15050000, 0, ID{Hash: checksumToBytes(0xf0afd0e3), Next: 18000000}},   // First Gray Glacier block
		{18000000, 0, ID{Hash: checksumToBytes(0x4fb8a872), Next: 1668000000}}, // First Merge Start block
		{20000000, 0, ID{Hash: checksumToBytes(0x4fb8a872), Next: 1668000000}}, // Last Merge Start block

> 		{20000000, 1668000000, ID{Hash: checksumToBytes(0xc1fdf181), Next: 0}}, // First Merge Start block

Reviewer 8: ```suggestion
		{20000000, 1668000000, ID{Hash: checksumToBytes(0xc1fdf181), Next: 0}}, // First Shanghai block
```

> 		{20000000, 2668000000, ID{Hash: checksumToBytes(0xc1fdf181), Next: 0}}, // Future Merge Start block

Reviewer 8: ```suggestion
		{20100000, 2669000000, ID{Hash: checksumToBytes(0xc1fdf181), Next: 0}}, // Future Shanghai block
```

	},
}
```

>  Here's a suite of tests of the different states a Mainnet node might be in and the different remote fork identifiers it might be required to validate and decide to accept or reject:

Reviewer 8: A summary of rejection cases are as follows
1. `Hash` of remote must be one of the hashes that the local node is aware of (past fork or future fork)
2. if `Hash` of remote is older than current `Hash` of local, remote `Next` must be equal to next fork (that exists in local)
3. if remote `Next` is smaller than local current (`BlockNumber` or `Timestamp`), local must have forked at that `Next` or reject remote.
Please correct me if i am wrong

	{20000000, 1668000001, ID{Hash: checksumToBytes(0x5cddc0e1), Next: 0}, ErrLocalIncompatibleOrStale},

	// Local is mainnet Withdrawals, and is aware of Petersburg. Remote announces Petersburg +
	// 0xffffffff. Local needs software update, reject.
	{20000000, 1668000001, ID{Hash: checksumToBytes(0x5cddc0e1), Next: 0}, ErrLocalIncompatibleOrStale},

	// Local is mainnet Withdrawals, remote is Rinkeby Petersburg.
	{20000000, 1668000001, ID{Hash: checksumToBytes(0xafec6b27), Next: 0}, ErrLocalIncompatibleOrStale},

	// Local is mainnet Withdrawals, far in the future. Remote announces Gopherium (non existing fork)
	// at some future block 88888888, for itself, but past block for local. Local is incompatible.
	//
	// This case detects non-upgraded nodes with majority hash power (typical Ropsten mess).
	{88888888, 1668000001, ID{Hash: checksumToBytes(0xf0afd0e3), Next: 88888888}, ErrRemoteStale},

	// Local is mainnet Withdrawals. Remote is in Byzantium, but announces Gopherium (non existing
	// fork) at block 7279999, before Petersburg. Local is incompatible.
	{20000000, 1668000001, ID{Hash: checksumToBytes(0xa00bc324), Next: 7279999}, ErrRemoteStale},
```

> 

Reviewer 4: ```suggestion
## Security Considerations 
No known security risks
```

  - The fork block numbers or timestamps are fed into the CRC32 checksum in ascending order.
  - If multiple forks are applied at the same block or time, the block number or timestamp is checksummed only once.
  - Block numbers are regarded as `uint64` integers, encoded in big endian format when checksumming.
  - Block timestamps are regarded as `uint64` integers, encoded in big endian format when checksumming.
  - If a chain is configured to start with a non-Frontier ruleset already in its genesis, that is NOT considered a fork.
- **`FORK_NEXT`**: Block number or timestamp (`uint64`) of the next upcoming fork, or `0` if no next fork is known.
  - Note that it is not important to distinguish between a timestamp or a block for `FROK_NEXT`.

A `FORK_HASH` for a timestamp based fork at `1668000000` on top of homestead would be:

- forkhash₁ = `0xcb37b2ee` (homestead+fictional fork) = `CRC32(<genesis-hash> || uint64(1150000) || uint64(1668000000))`

### Additional rules

The following additional rules are applied:

- Forks by timestamp MUST be scheduled at or after the forks by block (on mainnet as well as on private networks).

> - An implementation of forkid verification of remote peer needs to filter the incoming forkids first by block then by timestamp.

Reviewer 9: Does this mean that a client need to try comparing a block number of local head in the first place and then if it doesn't work fallback to the timestamp?

Author: Yes, we compare by block first and if the last block we have is past the remote block we fallback to the timestamp
